### PR TITLE
Add `OffscreenCanvas` support

### PR DESCRIPTION
**Pull request by daxpedda**
This is blocked on ~~https://github.com/rustwasm/wasm-bindgen/pull/3319~~ https://github.com/rustwasm/wasm-bindgen/pull/3416. Waiting for a new `web-sys` release.

Based on #76 (merged).

### CODE REVIEW

File: src/web.rs [hunk 1]
```diff
@@ -46,6 +47,18 @@ pub struct WebImpl {
     width: u32,
 }
 
+pub enum Canvas {
```
Comment (notgull) [inline src/web.rs:50]: Would be nice to document this enum. Also, does it need to be `pub`?

File: src/web.rs [hunk 2]
```diff
@@ -2,11 +2,15 @@
 
 #![allow(clippy::uninlined_format_args)]
 
+use js_sys::Object;
 use raw_window_handle::WebWindowHandle;
 use wasm_bindgen::JsCast;
+use wasm_bindgen::JsValue;
```
Comment (notgull) [inline src/web.rs:8]: Could you merge these two includes into one statement? Makes it cleaner.

File: src/web.rs [hunk 3]
```diff
@@ -105,6 +139,10 @@ impl WebImpl {
 pub trait SurfaceExtWeb: Sized {
     /// Creates a new instance of this struct, using the provided [`HtmlCanvasElement`].
     fn from_canvas(canvas: HtmlCanvasElement) -> Result<Self, SoftBufferError>;
+
+    /// Creates a new instance of this struct, using the provided [`HtmlCanvasElement`].
+    #[cfg(web_sys_unstable_apis)]
```
Comment (notgull) [inline src/web.rs:144]: Is this `cfg` stable? If so, the method documentation should mention it (and it should also be `doc(cfg(..))`)

File: src/web.rs [hunk 4]
```diff
@@ -46,6 +47,18 @@ pub struct WebImpl {
     width: u32,
 }
 
+pub enum Canvas {
+    Canvas {
+        canvas: HtmlCanvasElement,
+        ctx: CanvasRenderingContext2d,
+    },
+    #[cfg(web_sys_unstable_apis)]
```
Comment (daxpedda) [inline src/web.rs:55]: Not exactly ideal, but it works and doesn't really affect the public API if a user doesn't use `cfg(web_sys_unstable_apis)`.
As ugly as this is, it's how `wasm-bindgen` works.

We can remove this in the next `web-sys` release (https://github.com/rustwasm/wasm-bindgen/pull/3416) or wait another turn for it, WDYT @notgull?

**Review by notgull — CHANGES_REQUESTED** [review]

File: src/web.rs (hunk 4 above)
Comment (notgull) [reply src/web.rs:55]: I'd say to wait for it, since the next release shouldn't be that far away.

Comment (daxpedda) [reply src/web.rs:55]: Alright then.

File: src/web.rs (hunk 3 above)
Comment (daxpedda) [reply src/web.rs:144]: It's Rust stable, but it's not stable API in `web-sys`.
> (and it should also be `doc(cfg(..))`)

Just tried it, `feature(doc_auto_cfg)` already takes care of that.

File: src/web.rs (hunk 2 above)
Comment (daxpedda) [reply src/web.rs]: Done.

I guess we don't want to switch to nightly Rustfmt for [`[imports_granularity](https://rust-lang.github.io/rustfmt/?version=v1.5.1&search=#imports_granularity)`](https://rust-lang.github.io/rustfmt/?version=v1.5.1&search=#imports_granularity)?

File: src/web.rs (hunk 1 above)
Comment (daxpedda) [reply src/web.rs]: Done.

File: src/web.rs [hunk 5]
```diff
@@ -2,11 +2,14 @@
...
 
+use js_sys::Object;
 use raw_window_handle::WebWindowHandle;
-use wasm_bindgen::JsCast;
+use wasm_bindgen::{JsCast, JsValue};
 use web_sys::CanvasRenderingContext2d;
 use web_sys::HtmlCanvasElement;
 use web_sys::ImageData;
+use web_sys::OffscreenCanvas;
+use web_sys::OffscreenCanvasRenderingContext2d;
```
Comment (notgull) [inline src/web.rs:12]: These should be combined into one block.

**Review by notgull — CHANGES_REQUESTED** [review]

File: src/web.rs (hunk 5 above)
Comment (daxpedda) [reply src/web.rs]: Done.

**Review by notgull — APPROVED** [review]

**Comment by daxpedda** [comment]
Let's merge it after https://github.com/rust-windowing/softbuffer/pull/104, rebasing #103 is easier.

**Review by notgull — APPROVED** [review]
Looks good to me!

**Pull request merged**